internal/hcs/system.go

```diff
@@ -347,11 +462,68 @@ func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschem
 	if propertiesJSON == "" {
 		return nil, ErrUnexpectedValue
 	}
-	properties := &hcsschema.Properties{}
-	if err := json.Unmarshal([]byte(propertiesJSON), properties); err != nil {
+	props := &hcsschema.Properties{}
+	if err := json.Unmarshal([]byte(propertiesJSON), props); err != nil {
 		return nil, makeSystemError(computeSystem, operation, err, nil)
 	}
 
+	return props, nil
+}
+
+// PropertiesV2 returns the requested container properties targeting a V2 schema container.
+func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (_ *hcsschema.Properties, err error) {
+	computeSystem.handleLock.RLock()
+	defer computeSystem.handleLock.RUnlock()
+
+	// Let HCS tally up the total for VM based queries instead of querying ourselves.
+	if computeSystem.typ != "container" {
+		return computeSystem.hcsPropertiesV2Query(ctx, types)
+	}
+
+	// Define a starter Properties struct with the default fields returned from every
+	// query. Owner is only returned from Statistics but it's harmless to include.
+	properties := &hcsschema.Properties{
+		Id:            computeSystem.id,
+		SystemType:    computeSystem.typ,
+		RuntimeOsType: computeSystem.os,
+		Owner:         computeSystem.owner,
+	}
+
+	logEntry := log.G(ctx)
+	// First lets try and query ourselves without reaching to HCS. If any of the queries fail
+	// we'll take note and fallback to querying HCS for any of the failed types.
+	fallbackTypes, err := computeSystem.queryInProc(ctx, properties, types)
+	if err == nil && len(fallbackTypes) == 0 {
+		return properties, nil
+	} else if err != nil {
+		logEntry.WithError(err)
+		fallbackTypes = types
+	}
+
+	logEntry.WithFields(logrus.Fields{
+		logfields.ContainerID: computeSystem.id,
+		"propertyTypes":       fallbackTypes,
+	}).Warn("falling back to HCS for property type queries")
+
+	hcsProperties, err := computeSystem.hcsPropertiesV2Query(ctx, fallbackTypes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now loop through the fallback types and fill in any fields on the Properties struct. These are the set of
+	// queries that we have defined types for and that are handled in the V2 schema query code path in HCS.
+	for _, propType := range fallbackTypes {
+		switch propType {
+		case hcsschema.PTStatistics:
+			properties.Statistics = hcsProperties.Statistics
+		case hcsschema.PTProcessList:
+			properties.ProcessList = hcsProperties.ProcessList
+		case hcsschema.PTTerminateOnLastHandleClosed:
+			properties.TerminateOnLastHandleClosed = hcsProperties.TerminateOnLastHandleClosed
+		default:
```

Comment:
We should probably log here as well (or even error?) since right now if we got a type we don't support we'd just silently fail to return results.

Reply:
Actually a little worrisome how easily we could fail to return results here. I wonder if it's worth inverting so we copy the in-proc results into `hcsProperties` instead of the other way around?

Reply:
I believe HCS errors on any of the other types not defined here (besides memory for some reason, it just returns nothing for a container) so we wouldn't get to here, but it's better to be explicit.

Reply:
Returned an error here

Reply:
Oh, I didn't see your other comment, thanks Github -_-

Reply:
Changed to the approach in your second comment